### PR TITLE
mat: generalise SolveVec vector parameters

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

Updates #167.

### CODE REVIEW

File: mat/solve.go [hunk 1]
```diff
@@ -104,24 +104,37 @@ func (m *Dense) Solve(a, b Matrix) error {
...
 // SolveVec finds a minimum-norm solution to a system of linear equations defined
-// by the matrix a and the right-hand side vector b. If A is singular or
+// by the matrix a and the right-hand side column vector b. If A is singular or
 // near-singular, a Condition error is returned. Please see the documentation for
 // Dense.Solve for more information.
-func (v *VecDense) SolveVec(a Matrix, b *VecDense) error {
-	if v != b {
-		v.checkOverlap(b.mat)
+func (v *VecDense) SolveVec(a Matrix, b Vector) error {
+	if _, bc := b.Dims(); bc != 1 {
```
Comment (vladimir-ch) [inline mat/solve.go]: The other methods check the dims of `b` against the matrix. Should it be done here too?

File: mat/cholesky.go [hunk 2]
```diff
@@ -207,29 +207,34 @@ func (a *Cholesky) SolveChol(m *Dense, b *Cholesky) error {
 
 // SolveVec finds the vector v that solves A * v = b where A is represented
```
[the remark this answers is not in the document]
Comment (kortschak) [reply mat/cholesky.go]: Can we leave that for another PR?

File: mat/cholesky.go [hunk 3]
```diff
@@ -207,29 +207,34 @@ func (a *Cholesky) SolveChol(m *Dense, b *Cholesky) error {
...
 // by the Cholesky decomposition, placing the result in v.
-func (c *Cholesky) SolveVec(v, b *VecDense) error {
+func (c *Cholesky) SolveVec(v *VecDense, b Vector) error {
```
[the remark this answers is not in the document]
Comment (kortschak) [reply mat/cholesky.go]: Again, I think this is for another PR.

File: mat/lq.go [hunk 4]
```diff
@@ -209,17 +209,27 @@ func (lq *LQ) Solve(m *Dense, trans bool, b Matrix) error {
 
 // SolveVec finds a minimum-norm solution to a system of linear equations.
 // Please see LQ.Solve for the full documentation.
```
[the remark this answers is not in the document]
Comment (kortschak) [reply mat/lq.go]: Another PR.

#371

File: mat/solve.go (hunk 1 above)
Comment (kortschak) [reply mat/solve.go]: This method calls `(*Dense).Solve` and the dims are checked there. The `bc` check is special for this to ensure that the user has passed in a column vector.

File: mat/lq.go [hunk 5]
```diff
@@ -209,17 +209,27 @@ func (lq *LQ) Solve(m *Dense, trans bool, b Matrix) error {
...
-func (lq *LQ) SolveVec(v *VecDense, trans bool, b *VecDense) error {
-	if v != b {
-		v.checkOverlap(b.mat)
-	}
+func (lq *LQ) SolveVec(v *VecDense, trans bool, b Vector) error {
 	r, c := lq.lq.Dims()
+	if _, bc := b.Dims(); bc != 1 {
```
[the remark this answers is not in the document]
Comment (kortschak) [reply mat/lq.go]: Checked in `(*LQ).Solve`.

File: mat/qr.go [hunk 6]
```diff
@@ -203,19 +203,31 @@ func (qr *QR) Solve(m *Dense, trans bool, b Matrix) error {
...
+// SolveVec finds a minimum-norm solution to a system of linear equations,
+//  Ax = b.
 // Please see QR.Solve for the full documentation.
-func (qr *QR) SolveVec(v *VecDense, trans bool, b *VecDense) error {
-	if v != b {
-		v.checkOverlap(b.mat)
-	}
+func (qr *QR) SolveVec(v *VecDense, trans bool, b Vector) error {
 	r, c := qr.qr.Dims()
+	if _, bc := b.Dims(); bc != 1 {
```
[the remark this answers is not in the document]
Comment (kortschak) [reply mat/qr.go]: Checked in `(*QR).Solve`.

**Pull request merged**